src/meta_schedule/profiler.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "./utils.h"
+
+namespace tvm {
+namespace meta_schedule {
+
+/**************** Context Manager ****************/
+
+struct ScopedTimer {
+  std::function<void()> func;
+
+  explicit ScopedTimer(std::function<void()> func) : func(func) {}
+  ~ScopedTimer() {
+    LOG(INFO) << "destructed";
+    func();
+  }
+};
+
+struct ProfilerThreadLocalEntry {
+  Optional<Profiler> ctx;
+};
+
+class ProfilerInternal {
+ public:
+  static void EnterScope(Profiler ctx) { ctx.EnterWithScope(); }
+  static void ExitScope(Profiler ctx) { ctx.ExitWithScope(); }
+};
+
+using ProfilerThreadLocalStore = dmlc::ThreadLocalStore<ProfilerThreadLocalEntry>;
+
+void Profiler::EnterWithScope() {
+  Optional<Profiler>& ctx = ProfilerThreadLocalStore::Get()->ctx;
+  CHECK(!ctx.defined()) << "ValueError: Nested Profiler context managers are not allowed";
+  ctx = *this;
+}
+
+void Profiler::ExitWithScope() {
+  Optional<Profiler>& ctx = ProfilerThreadLocalStore::Get()->ctx;
+  ICHECK(ctx.defined());
+  ctx = NullOpt;
+}
+
+/**************** Profiler ****************/
+
+Profiler::Profiler() {
+  ObjectPtr<ProfilerNode> n = make_object<ProfilerNode>();
+  data_ = n;
+}
+
+ScopedTimer ProfilerNode::TimeIt(String name) {
+  return ScopedTimer([this, name, tick = std::chrono::high_resolution_clock::now()]() -> void {
+    double duration = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                          std::chrono::high_resolution_clock::now() - tick)
+                          .count() /
+                      1e9 / 60;
+    Map<String, FloatImm> stats = this->stats;
```

Comment:
note: Map has copy-on-write semantics 

```suggestion
    Map<String, FloatImm>& stats = this->stats;
```

Reply:
Noted with thanks!